### ISSUE TITLE
pq: unsupported comparison operator: <string> = ANY <uuid[]>

### ISSUE THREAD

**Issue by derkan** [cockroachdb/cockroach#24175]
Is this a question, feature request, or bug report?

> This is a bug report.

**BUG REPORT**
     grep -F '[config]' cockroach-data/logs/cockroach.log
```
I180320 16:52:22.923088 518 util/log/clog.go:938  [config] file created at: 2018/03/20 16:52:22
I180320 16:52:22.923088 518 util/log/clog.go:938  [config] running on machine: db1
I180320 16:52:22.923088 518 util/log/clog.go:938  [config] binary: CockroachDB CCL v1.1.6 (linux amd64, built 2018/03/12 17:58:05, go1.8.3)
```

2. Please describe the issue you observed:

This SQL works as expected:
```sql
SELECT '3ae3560e-d771-4b63-affb-47e8d7853680'::UUID = ANY ('{3ae3560e-d771-4b63-affb-47e8d7853680}'::UUID[])as a;
+------+
|  a   |
+------+
| true |
+------+
(1 row)

```
But when checking against table column with type:
```sql
shared_users      UUID[]
```
with this sql, it fails:
```sql
SELECT * FROM documents WHERE '3ae3560e-d771-4b63-affb-47e8d7853680'::UUID = ANY (documents.shared_users);
pq: unsupported comparison operator: <string> = ANY <uuid[]>
```

A workaround is to wrap `UUID` value in a inner `SELECT`:
```sql
SELECT shared_users FROM documents WHERE (SELECT '3ae3560e-d771-4b63-affb-47e8d7853680'::UUID) = ANY (documents.shared_users);
+--------------+
| shared_users |
+--------------+
+--------------+
(0 rows)


```

I think there is a problem while checking type matching for `ANY` operator on arrays.

**Comment by 0xbaadf00d** [comment]
Hello, seems fixed on the dev version 2.0-beta. Could try ? https://www.cockroachlabs.com/docs/dev/install-cockroachdb.html

**Comment by derkan** [comment]
Tested on build `CCL v2.0-beta.20180319 @ 2018/03/19 18:57:22 (go1.10)` 
Same problem exists with different error message:

```sql
create database testdb;
use testdb;
create table documents ( shared_users      UUID[] ) ;
-- this is OK:
SELECT shared_users FROM documents WHERE (SELECT '3ae3560e-d771-4b63-affb-47e8d7853680'::UUID) = ANY (documents.shared_users);
+--------------+
| shared_users |
+--------------+
+--------------+
(0 rows)

-- this has error:
SELECT * FROM documents WHERE '3ae3560e-d771-4b63-affb-47e8d7853680'::UUID = ANY (documents.shared_users);
pq: '3ae3560e-d771-4b63-affb-47e8d7853680' = SOME "$0": unsupported comparison operator: <string> = SOME <uuid[]>
```

**Comment by jordanlewis** [comment]
Thanks for the report - there's definitely something wrong going on here. I can reproduce on the latest `master` as well

**Comment by knz** [comment]
@justinj can you check whether the optinizer already supports this? If it does, perhaps  add a test for this caase there, and we'll close this  issue as addressed in 2.1.

**Comment by andy-kimball** [comment]
This appears to be fixed both with the heuristic planner and the cost-based optimizer. I added a regression case as well.
